Design of data preprocessing in AutoML.md

```diff
@@ -0,0 +1,93 @@
+
+Not sure which approach is better but probably we can avoid introduction of the PipelineModelBuilder class. But it will increase number of responsibilities of the ModelBuilder class.
+
+#### Pros:
+- We do not hide any functionality behind PipelineModel wrapper
+- All models could be equally seen as pipelines ( which could also be a cons.)
+
+#### Cons:
+- conceptually abusing notion of the Model. Model ( what is also known as Transformer in other frameworks) is supposed to be only trained on data and after that be able to transform similar data. Nothing about data preprocessing is expected here. See Concerns section for more details.
+- new functionality will be exposed not only to AutoML scope. All algos will be "aware" about data preprocessing from now on. With first approach we would just add notion of ability to be a Proxy model and not all the models will have to realise this interface. Though, absense of the preprocessing models could be seen as "no-impact" on Model, still from code perspective there will a bigger impact. Hopefully preprocessing functionality will be generic enough without depenedencies on a particular types of the models.
+- increased number of responsibilities of the ModelBuilder and Model, which could be otherwise be located in other classes.
+
+
+![Integration preprocessing functionality into a Model class](no_pipelinemodel.png)
+
+
+#### General concerns:
+Wanted one more time to draw our attention to the fact how we handle cornerstone abstractions in our project. And let's say we are not going to use Pipelines.
+
+For comparison I will be using SparkML framework:
+
+![SparkML vs H2O.ai](sparkml_vs_h2oai.png)
+
+ref. https://www.youtube.com/watch?v=jEyahxFp3ak
+
+Model is being overflowed with responsibilities imho. Transformation logic is being spread over many places in different forms.
+
+### Summary
+- we tend to keep adding functionality into Model class. We might be going to add pipeline related aspect into Model
+- lack of symmetry between production and non-production code ( Model vs MojoPipeline )
+- both approaches can solve the problem. Second approach will be smoother from user's perspective at least in Java API as there will be no need to pull out main model. For clients I think we can flatten representation automatically.
+- Some functionality that could be represented via existing approach of Models(transformer) is hardcoded. Lack of consistency. ( not too much related to a current task)
+
+
+Given current state of h2o-3's design and given we are ready to put responsibility of data preprocessing onto a Model class I think that second approach should be fine.
+But ideally I would want to see
+ - a PipelineModel related classes to be responsible for pipelining/data preprocessing
+ - and hardcoded transformations to be represented as models. ( not too much related to a current task)
+
+### Other approaches to consider ???
+
```

Comment:
- H2OAssembly - largely abandoned at this point but needs to be mentioned becae "Assemblies are essentially Pipelines"

Reply:
Yeah, we also mentioned it in our discussions.
Agree that it should be mentioned, with its limitations: abandoned, supports only `Transformer`s, doesn't provide any single model feature.

Reply:
@michalkurka yes, thank you. @sebhrusen told me about existence of `Assembly` in our automl discussion. I forgot to add it to the list.